linera-core/src/chain_worker/state.rs

```diff
@@ -1261,14 +1265,49 @@ where
         chain_id = %self.chain_id(),
         query_application_id = %query.application_id()
     ))]
-    async fn query_application(&mut self, query: Query) -> Result<QueryOutcome, WorkerError> {
+    pub(super) async fn query_application(
+        &mut self,
+        query: Query,
+        block_hash: Option<CryptoHash>,
+    ) -> Result<QueryOutcome, WorkerError> {
         self.initialize_and_save_if_needed().await?;
         let local_time = self.storage.clock().current_time();
-        let outcome = self
-            .chain
-            .query_application(local_time, query, self.service_runtime_endpoint.as_mut())
-            .await?;
-        Ok(outcome)
+        if let Some(requested_block) = block_hash {
```

Comment:
You could make this a `let … else` and return early in the simple case (currently `else`).